### PR TITLE
DAOS-10488 test: Add dfuse unit test code and call it from ftest.

### PR DESCRIPTION
**Pull request by ashleypittman**
Quick-Functional: true
Test-tag: dfuse_test

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by brianjmurrell — COMMENTED** [review]
Maybe this isn't ready for review yet and you just forgot to open it as a draft.  If so, you can switch it to a draft to avoid getting any more watcher/codeowner reviews.

In any case, don't forget to bump `Release:`, make a `%changelog` entry and the corresponding `debian/` changes.

File: src/tests/ftest/daos_test/dfuse.py
Comment (ashleypittman) [reply src/tests/ftest/daos_test/dfuse.py:3]: This is a PR from 2021.  I should probably have set it to draft before updating it.

**Comment by ashleypittman** [comment]
I think this is complete, stable, python2/3 safe now and is correctly working with Test-tag: so is ready for review/landing.

@brianjmurrell is there anything else needed with the rpm changes?  Do I need to add a changelog and version numbers?  If so I'd like to get at least one round of reviews in as I know making that change can be conflict-heavy.

**Comment by ashleypittman** [comment]
I think this is complete, stable, python2/3 safe now and is correctly working with Test-tag: so is ready for review/landing.

@brianjmurrell is there anything else needed with the rpm changes?  Do I need to add a changelog and version numbers?  If so I'd like to get at least one round of reviews in as I know making that change can be conflict-heavy.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
https://github.com/daos-stack/daos/pull/7286#pullrequestreview-911555850 still needs to be addressed.

File: src/tests/ftest/daos_test/dfuse.yaml
Comment (daltonbohning) [inline src/tests/ftest/daos_test/dfuse.yaml:33]: Optional - Just FYI that you can do something like this for shared config values
```suggestion
dfuse:
  mount_dir: "/tmp/daos_dfuse/"
  io_mode: !mux
    writeback:
      # File metadata is incorrect after write with writeback caching and IL.
      !filter-only : "/run/intercept/off"  # yamllint disable-line rule:colons
      name: "writeback"
    writethrough:
      disable_wb_cache: true
      name: "writethrough"
    metadata:
      disable_wb_cache: true
      name: "metadata"
    nocache:
      disable_wb_cache: true
      disable_caching: true
      name: "nocache"
```

Comment (ashleypittman) [reply src/tests/ftest/daos_test/dfuse.yaml:33]: I might experiment with this later once I've got my head around it but I'm blocked on these tests landing and CI turnaround is slow so I won't make changes now.

**Review by daltonbohning — APPROVED** [review]
Overall LGTM but unsure whether the launch.py changes require harness tests to be ran

**Comment by ashleypittman** [comment]
> Overall LGTM but unsure whether the launch.py changes require harness tests to be ran

Is that just a Test-tag or something else?  I think they're fairly safe as it's just the yaml loading.  The "new" yaml tags won't be included in the returned data structure but the fact that I needed to add the code to add the tags says there's no other tags in use anyway.

**Comment by daltonbohning** [comment]
> Is that just a Test-tag or something else? I think they're fairly safe as it's just the yaml loading. The "new" yaml tags won't be included in the returned data structure but the fact that I needed to add the code to add the tags says there's no other tags in use anyway.

These: https://github.com/daos-stack/daos/tree/master/src/tests/ftest/harness
But I'm not familiar.

> the fact that I needed to add the code to add the tags

What do you mean? Which tags?

**Comment by ashleypittman** [comment]
> > Is that just a Test-tag or something else? I think they're fairly safe as it's just the yaml loading. The "new" yaml tags won't be included in the returned data structure but the fact that I needed to add the code to add the tags says there's no other tags in use anyway.
> 
> These: https://github.com/daos-stack/daos/tree/master/src/tests/ftest/harness But I'm not familiar.
> 
> > the fact that I needed to add the code to add the tags
> 
> What do you mean? Which tags?

Yaml tags, the ! notation in the yaml files.

**Comment by ashleypittman** [comment]
> > Yaml tags, the ! notation in the yaml files.
> 
> That much I thought already worked. We use `!mux` all over. There are several cart tests with `!filter`, though they seem to be commented out:
> 
> https://github.com/daos-stack/daos/blob/9c4ac298bf33dfdd4f091ccc2b309796141112cd/src/tests/ftest/cart/corpc/corpc_five_node.yaml#L6-L7

As per this PR it changes launch.py from using a string replace to remove the !mux to using json parser handlers to continue over !mux tags and to drop all others, so it did already work but only for those tags and in a different way.

**Review by jolivier23 — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/launch.py
Comment (phender) [reply src/tests/ftest/launch.py:1410]: The only concern would be making sure this change doesn't cause a problem with any existing yaml files.  Running the `pr` tagged tests will cover most of the cases, but we would want to verify compliance with a test that uses `!mux` in the `test_servers:` assignment, like https://github.com/daos-stack/daos/blob/master/src/tests/ftest/erasurecode/online_rebuild_mdtest.yaml.  The `test_ec_online_rebuild_mdtest` has a `pr` tag so this should be covered.  

Running the following command manually should verify this will not break any pr, daily, or weekly testing:
```
./launch.py -l -m -d -n auto -ts wolf-[A-B] pr daily_regression full_regression
```
(`wolf-[A-B]` does not need to be a large list of host - 2 hosts with at least 1 NVMe each should suffice.)

**Review by phender — APPROVED** [review]
The ftest changes look good but I would highly recommend running this command manually to verify the `launch.py` changes will not break pr, daily, and weekly testing:
```
./launch.py -l -m -d -n auto -ts wolf-[A-B] pr daily_regression full_regression
```

File: src/tests/ftest/launch.py
Comment (ashleypittman) [reply src/tests/ftest/launch.py:1410]: Seems to work correctly from all my testing, and I added a print to the return of this function to be sure.  This is only loading the yaml files into a python datatype so is a fairly self-contained change.

Here are the results for that test, it has three variants as expected https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-7286/52/testReport/FTEST_erasurecode/EcodOnlineRebuildMdtest/

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**
